### ISSUE TITLE
plot_images

### ISSUE THREAD

**Issue by Q-qqq** [ultralytics/ultralytics#22000]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

# plotting.plot_images
我发现train_batch图像的分割标签出现错位现象：矩形框标签标注正确，但分割标签不对应矩形框，跑到下一个图像去了，然后发现plot_images函数中if idx.shape[0] == masks.shape[0]:  # overlap_masks=False，该条件运行时判断为True，而我的overlap_masks设置为默认值True，因为batch_idx为0、1、3、4、5、6、6、7，数量同样为8，但不满足0-8的顺序排序，从而出现错误判断

### Environment

none

### Minimal Reproducible Example

# Plot masks
                if len(masks):
                    if idx.shape[0] == masks.shape[0]:  # overlap_masks=False
                        image_masks = masks[idx]
                    else:  # overlap_masks=True
                        image_masks = masks[[i]]  # (1, 640, 640)
                        nl = idx.sum()
                        index = np.arange(nl).reshape((nl, 1, 1)) + 1
                        image_masks = np.repeat(image_masks, nl, axis=0)
                        image_masks = np.where(image_masks == index, 1.0, 0.0)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Q-qqq, thanks for opening this issue and for the clear description 🚀 This is an automated response to help get things moving quickly—an Ultralytics engineer will follow up shortly.

For new users, we recommend checking the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples and answers to common questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us debug. Since your report involves `plotting.plot_images` mask/box misalignment and `overlap_masks` behavior, please include:
- A minimal script showing how you call `plot_images` and how `overlap_masks` is set
- A small sample of your labels (1–2 images) that reproduces the issue
- Exact `ultralytics` version and `pip list | grep ultralytics` output
- Console logs and a screenshot of the plotted batch illustrating the mismatch
- Any custom dataloader/augmentation settings that may affect `batch_idx` or masks
- Whether this occurs during training, validation, or standalone plotting

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for the detailed context around `batch_idx` ordering and `masks` alignment in `plot_images`—with an MRE, we can reproduce and investigate quickly 🔍

**Comment by Y-T-G** [comment]
Is the issue similar to this?

https://github.com/ultralytics/ultralytics/issues/18832#issuecomment-2674170867

**Comment by Q-qqq** [comment]
Yes

**Comment by glenn-jocher** [comment]
Thanks for confirming. You’re right—the heuristic in `plot_images` that checks `idx.shape[0] == masks.shape[0]` can misclassify overlap masks when `batch_idx` is non‑contiguous and `len(labels) == bs`. A safer check is to key off batch size instead:

```python
# inside ultralytics/utils/plotting.py -> plot_images()
# after bs, h, w are defined and before selecting image_masks
if len(masks):
    overlap = masks.ndim == 3 and masks.shape[0] == bs  # detect overlap-masks per image
    if not overlap:  # per-instance masks
        image_masks = masks[idx]
    else:  # overlap-masks=True
        image_masks = masks[[i]]  # (1, H, W)
        nl = int(idx.sum())
        index = np.arange(nl).reshape((nl, 1, 1)) + 1
        image_masks = np.repeat(image_masks, nl, axis=0)
        image_masks = np.where(image_masks == index, 1.0, 0.0)
```

This removes the assumption that `batch_idx` is sequential and should align masks correctly. For reference, see the `plot_images` API and behavior in the docs, and how overlap masks are produced via `polygons2masks_overlap` in the dataloader: [plot_images reference](https://docs.ultralytics.com/reference/utils/plotting/#plot_images), [polygons2masks_overlap](https://docs.ultralytics.com/reference/data/utils/#polygons2masks_overlap).

If this still misaligns on your data after applying the change, please share a minimal batch dump of shapes for `cls`, `bboxes`, `masks`, and `batch_idx` plus the full text traceback, or feel free to open a small PR with this guard and a unit test—happy to review.

**Comment by Q-qqq** [comment]
Regardless of whether overlap_mask is True or False, the ndim of masks is 3, and it cannot be guaranteed that masks.shape[0] == bs. The order of idx is sequentially arranged, and when masks.shape[0] == bs == 8, it may be that idx == bs == 8 with the order of idx being 1, 3, 4, 5, 6, 6, 7.

**Comment by Y-T-G** [comment]
The problem is if the total number of objects in the batch becomes equal to batch size for plotting, we get ambiguity. We expect the total number of objects to be equal to total number of masks for `overlap_masks=True`, but it's possible to have edge case where number of objects is equal to number of masks outside of that, like in your case. But the misalignment occurs, if some of your images have no objects and another has more than 1, and the total still becomes equal to total number of masks or batch size. In your case, there are no objects in the third image, while the 7th image has two objects, so total is still 8.

**Comment by UltralyticsAssistant** [comment]
Good news — a potential fix for this issue has been merged in the pull request Fix mask plotting when number of objects equal to number of images in a batch.

Highlights that may resolve this:
- Correctly treats masks as binary only when the maximum value is 1 or less, preventing misinterpretation of instance-labeled masks.
- Refactors overlap handling using NumPy broadcasting instead of repeating arrays, improving both correctness and performance.
- Simplifies and corrects instance ID matching with clean index generation.
- Ensures output masks are explicitly float32 for consistent plotting downstream.

Credit to the author Y-T-G with contributions from Laughing-q for the thoughtful improvements.

How to test the fix:
- Install the latest from GitHub main:
  pip install -U git+https://github.com/ultralytics/ultralytics.git@main
- Or update to the latest PyPI release once it’s published:
  pip install -U ultralytics
- Example quick check for segmentation plotting:
  ```python
  from ultralytics import YOLO

  model = YOLO("yolo11n-seg.pt")  # any YOLO11 segmentation model
  results = model("path/to/images_or_dir", save=True)  # check saved images for correct mask overlays
  ```

If possible, try reproducing your original scenario (where the number of objects equals the number of images in the batch) and confirm whether mask overlays now render correctly.

Feedback on whether this resolves the issue would be greatly appreciated. Thanks for reporting it, and please share any remaining details if the problem persists.

**Comment by Q-qqq** [comment]
Yes, that's the situation.

**Comment by glenn-jocher** [comment]
Thanks for confirming—that edge case is now fixed in `plot_images` on main. Please update and re-test; to verify you’re on the patched build, run:

```bash
yolo checks
```

If the misalignment persists, open a new issue with a minimal batch dump of shapes for `cls`, `bboxes`, `masks`, and `batch_idx` (text preferred) so we can reproduce.
